apps/emqx_message_transformation/src/emqx_message_transformation.erl

```diff
@@ -0,0 +1,820 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+-module(emqx_message_transformation).
+
+-include_lib("snabbkaffe/include/trace.hrl").
+-include_lib("emqx_utils/include/emqx_message.hrl").
+-include_lib("emqx/include/emqx_hooks.hrl").
+-include_lib("emqx/include/logger.hrl").
+
+%% API
+-export([
+    add_handler/0,
+    remove_handler/0,
+
+    load/0,
+    unload/0,
+
+    list/0,
+    reorder/1,
+    lookup/1,
+    insert/1,
+    update/1,
+    delete/1
+]).
+
+%% `emqx_hooks' API
+-export([
+    register_hooks/0,
+    unregister_hooks/0,
+
+    on_message_publish/1
+]).
+
+%% `emqx_config_handler' API
+-export([pre_config_update/3, post_config_update/5]).
+
+%% `emqx_config_backup' API
+-behaviour(emqx_config_backup).
+-export([import_config/1]).
+
+%%------------------------------------------------------------------------------
+%% Type declarations
+%%------------------------------------------------------------------------------
+
+-define(TRACE_TAG, "MESSAGE_TRANSFORMATION").
+-define(CONF_ROOT, message_transformation).
+-define(CONF_ROOT_BIN, <<"message_transformation">>).
+-define(TRANSFORMATIONS_CONF_PATH, [?CONF_ROOT, transformations]).
+
+-type transformation_name() :: binary().
+%% TODO
```

Comment:
Nit: m.b. add some description of what is yet to be implemented

Reply:
It's a todo for a more detailed typespec :sweat_smile: :see_no_evil:

Reply:
:heavy_check_mark: